prusti-contracts/src/ghost.rs

```diff
@@ -0,0 +1,167 @@
+use std::ops::*;
+use core::marker::PhantomData;
+
+pub unsafe auto trait Ghost {}
+
+macro_rules! implement_ghost_type {
+    ($ghost_type: ident) => {
+        #[derive(PartialEq, Eq)]
+        pub struct $ghost_type;
+        unsafe impl Ghost for $ghost_type{}
+    };
+}
+
+macro_rules! implement_ghost_type_generic {
+    ($ghost_type: ident) => {
+        pub struct $ghost_type<T: Ghost> {
+            _type: PhantomData<T>
+        }
+        impl <T: Ghost> $ghost_type<T> {
+            /// Constructor for generic ghost types.
+            /// #Examples
+            ///  ```rust
+            ///  use prusti_contracts::{GhostInt, GhostSeq};
+            ///  let seq_inst: GhostSeq<GhostInt> = GhostSeq::new();
+            ///  ```
+            pub fn new() -> Self {
+                $ghost_type {
+                    _type: PhantomData,
+                }
+            }
+        }
+        unsafe impl<T: Ghost> Ghost for $ghost_type<T> {}
+    }
+}
+
+// Ghost variant of Viper type, Int
+implement_ghost_type!(GhostInt);
+// wrappers around standard operations on GhostInt
+impl Add for GhostInt {
+    type Output = Self;
+    fn add(self, other: GhostInt) -> Self::Output {
+        GhostInt
+    }
+}
+
+impl Sub for GhostInt {
+    type Output = Self;
+    fn sub(self, other: GhostInt) -> Self::Output {
+        GhostInt
+    }
+}
+
+impl GhostInt {
+    pub fn new(val: i32) -> Self {
+        GhostInt
+    }
+}
+
+// Ghost variant of Viper type, Bool
+implement_ghost_type!(GhostBool);
+// wrappers around standard operations on GhostBool
+impl BitAnd for GhostBool {
+    type Output = Self;
+    fn bitand(self, rhs: Self) -> Self::Output {
+        GhostBool
+    }
+}
+
+impl BitOr for GhostBool {
+    type Output = Self;
+    fn bitor(self, rhs: Self) -> Self::Output {
+        GhostBool
+    }
+}
+
+impl Not for GhostBool {
+    type Output = Self;
+    fn not(self) -> Self::Output {
+        GhostBool
+    }
+}
+
+// Ghost variant of Viper type, Seq[T]
+implement_ghost_type_generic!(GhostSeq);
+// wrappers around standard operations on GhostSeq
+impl<T: Ghost> GhostSeq<T> {
```

Comment:
Is there a method to retrieve the i'th element of a sequence?

Reply:
I've modified this in a later commit. So now this method is supported.